File: solutions/java/src/pubsubsystem/Topic.java
package pubsubsystem;

import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

public class Topic {
    private final String name;
    private final Set<Subscriber> subscribers = new CopyOnWriteArraySet<>();

    public Topic(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public void addSubscriber(Subscriber subscriber) {
        subscribers.add(subscriber);
    }

    public void removeSubscriber(Subscriber subscriber) {
        subscribers.remove(subscriber);
    }

    public void publish(Message message) {
        for (Subscriber subscriber : subscribers) {
            subscriber.onMessage(message);
        }
    }
}

File: solutions/csharp/pubsubsystem/Topic.cs
using System.Collections.Concurrent;
using System.Collections.Generic;

namespace PubSubSystem
{
    public class Topic
    {
        public string Name { get; }
        private readonly ConcurrentBag<ISubscriber> _subscribers;

        public Topic(string name)
        {
            Name = name;
            _subscribers = new ConcurrentBag<ISubscriber>();
        }

        public void AddSubscriber(ISubscriber subscriber)
        {
            _subscribers.Add(subscriber);
        }

        public void RemoveSubscriber(ISubscriber subscriber)
        {
            _subscribers.TryTake(out subscriber);
        }

        public void Publish(Message message)
        {
            foreach (var subscriber in _subscribers)
            {
                subscriber.OnMessage(message);
            }
        }
    }
}